Implementation of unsharp masking filter.

dvolgyes opened the pull request:
## Description
Implementation of unsharp masking filter.

This PR mostly solves #2477, except that I disagree with the multichannel case.
Color images could be in arbitrary color space. If the user wants to use a specific color space,
then he/she should convert the data there first, and process it afterwards.
HSV could be a nice color space, but also YCbCr is a nice one, and a lot of people would argue that
this space should be used because it is more perceptually uniform than most other color spaces.

Higher dimensional array also could be processed, as well as different number of color-like channels, e.g. RGBA, and so on. So this PR implements classic unsharp masking for one or more dimensional datasets, with or without colors using arbitrary input types. The will have the same type as the input.

A similar PR was opened ( #2642 ) but the implementation there is based on an adaptive variation,
which is nice but not as straightforward as the original one. That PR should be renamed to adaptive_unsharp_masking. 

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]


## References

It implements #2477, and related to #2642.

The implementation is based on this books:
Jayaraman, S., Esakkirajan, S., & Veerakima, T. "Digital image processing", (2009), ed iii., page 274

Wikipedia article: [https://en.wikipedia.org/wiki/Unsharp_masking](https://en.wikipedia.org/wiki/Unsharp_masking)

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

pep8speaks commented:
Hello @dvolgyes! Thanks for submitting the PR.

 - In the file [`doc/examples/filters/plot_unsharp_masking.py`](https://github.com/scikit-image/scikit-image/blob/9cd579832a4248fc1f29c041217e4997493a9a70/doc/examples/filters/plot_unsharp_masking.py), following are the PEP8 issues :

> [Line 44:80](https://github.com/scikit-image/scikit-image/blob/9cd579832a4248fc1f29c041217e4997493a9a70/doc/examples/filters/plot_unsharp_masking.py#L44): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)





 - In the file [`skimage/filters/unsharp_masking.py`](https://github.com/scikit-image/scikit-image/blob/9cd579832a4248fc1f29c041217e4997493a9a70/skimage/filters/unsharp_masking.py), following are the PEP8 issues :

> [Line 98:80](https://github.com/scikit-image/scikit-image/blob/9cd579832a4248fc1f29c041217e4997493a9a70/skimage/filters/unsharp_masking.py#L98): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)

dvolgyes commented:
The CI tests fail but I am not quite sure why. It seems to me that some other parts of skimage fails, not mine, but if you can show me where my code fails, I will fix it.

dvolgyes commented on skimage/filters/tests/test_unsharp_masking.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def test_unsharp_masking_output_type_with_different_array_shapes():
+    for shape, multichannel in [((101,), False),
```
Indeed.

dvolgyes commented on skimage/filters/tests/test_unsharp_masking.py:
```diff
@@ -0,0 +1,49 @@
+                assert output.shape == shape
+
+
+if __name__ == "__main__":
```
OK.

dvolgyes commented on skimage/filters/unsharp_masking.py:
```diff
@@ -0,0 +1,108 @@
+"""
+Unsharp masking is an image sharpening technique.
```
Fine.

dvolgyes commented on skimage/filters/unsharp_masking.py:
```diff
@@ -0,0 +1,108 @@
+See more details and history on Wikipedia site[2]_
+of unsharp masking.
+
+.. [1]  Jayaraman, S., Esakkirajan, S., & Veerakima, T.
```
OK.

dvolgyes commented on skimage/filters/unsharp_masking.py:
```diff
@@ -0,0 +1,108 @@
+
+    Parameters
+    ----------
+    data : array_like
```
OK.

dvolgyes commented on skimage/filters/unsharp_masking.py:
```diff
@@ -0,0 +1,108 @@
+    ----------
+    data : array_like
+        Image to be filtered.
+
```
OK.

dvolgyes commented on skimage/filters/unsharp_masking.py:
```diff
@@ -0,0 +1,108 @@
+    data : array_like
+        Image to be filtered.
+
+    radius : float or tuple of floats. If a single radius is given,
```
OK

dvolgyes commented on skimage/filters/unsharp_masking.py:
```diff
@@ -0,0 +1,108 @@
+        The factor could be 0 or negative, but it is usually a
+        small positive number, e.g. 1.0.
+
+    multichannel : bool, default False
```
How do you indicate the value if it is optional?
multichannel : bool, optional (False)?
radius: float, optional (2.0)?

soupault replied:
@dvolgyes Just `multichannel : bool, optional` or `radius: float, optional`. The default value is visible in function signature (see http://scikit-image.org/docs/dev/api/skimage.filters.html#skimage.filters.gaussian for example).

soupault commented:
@dvolgyes could you enable access to your branch for maintainers? (see https://help.github.com/assets/images/help/pull_requests/allow-maintainers-to-make-edits-sidebar-checkbox.png)

soupault commented:
Testing with the current matrix takes 24 seconds on a moderate laptop.
Gallery example:
![sphx_glr_plot_unsharp_mask_001](https://user-images.githubusercontent.com/1315589/29747977-3b090416-8b14-11e7-97c4-37585c3abb8f.png)

dvolgyes commented:
I think I enabled it from the beginning.
At least it is enabled at this moment.

On Sun, Aug 27, 2017 at 9:40 AM, Egor Panfilov <notifications@github.com>
wrote:

> @dvolgyes <https://github.com/dvolgyes> could you enable access to your
> branch for maintainers? (see https://help.github.com/
> assets/images/help/pull_requests/allow-maintainers-to-
> make-edits-sidebar-checkbox.png)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-image/scikit-image/pull/2772#issuecomment-325182519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ-WBtLF7bCbQpiqWxD_lv9xPRgN-96ks5scR1xgaJpZM4PBcDN>
> .
>

dvolgyes commented:
Test:
it takes a while, and it doesn't cover all the possible cases, but I
thought it is a reasonable well covered case:
-odd and even number of pixels
-different types
-negative values
etc.
It depends on the taste and testing practice, so feel free to modify it, if
you have any better suggestion.

The gallery shows a few options but it could be refined if needed. E.g.
color image, etc.


On Sun, Aug 27, 2017 at 9:41 AM, Egor Panfilov <notifications@github.com>
wrote:

> Testing with the current matrix takes 24 seconds on a moderate laptop.
> Gallery example:
> [image: sphx_glr_plot_unsharp_mask_001]
> <https://user-images.githubusercontent.com/1315589/29747977-3b090416-8b14-11e7-97c4-37585c3abb8f.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-image/scikit-image/pull/2772#issuecomment-325182554>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ-WGH_60PaPbPuRUYqW0hUis8w8ZO5ks5scR2ggaJpZM4PBcDN>
> .
>

dvolgyes commented:
I am sorry for the long response time, this weekend is a bit busy for me.
Thank you all for the improvements. 
The [-1,1] float range for images a bit strange choice for me, but I understand that scikit uses this convention. Unsharp masking is independent from scaling, so it should work nicely in this range too.
If there is anything I can contribute, let me know, but I will not have time to actually implement it until Monday noon/afternoon.

soupault commented:
>I am sorry for the long response time, this weekend is a bit busy for me.
Thank you all for the improvements.

No need to sorry, everything was fine. I'd appreciate if you could look through my changes to see more conventions we use in `skimage`.

>The [-1,1] float range for images a bit strange choice for me, but I understand that scikit uses this convention.

I understand your point. I think this representation became very popular recently with the rise of deep learning algorithms where the data is typically centered around 0 and normalized.

stefanv commented on doc/examples/filters/plot_unsharp_mask.py:
```diff
@@ -0,0 +1,46 @@
+Unsharp masking
+===============
+
+Unsharp masking is a linear image processing technique which sharpens
```
What does 'linear' mean in this context?

dvolgyes replied:
It is linear in mathematical sense. If you ignore the rounding errors and clipping due to overflow/underflow, then:
```USM(A+B) = USM(A)+USM(B)```

```USM(A*alpha) = USM(A)*alpha```

It is basically a convolution with a special kernel.

stefanv commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,104 @@
+    result = image + (image - blurred) * amount
+
+    # Enforce image dtype limits
+    if image.dtype.kind == 'f':
```
Should probably be `np.issubdtype(image.dtype, float)`.

stefanv replied:
Ping on this one.

soupault replied:
@stefanv could you describe the difference between these two alternatives?

stefanv commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,104 @@
+                image[..., channel], radius, amount)
+        return result
+    else:
+        return _unsharp_mask_single_channel(image, radius, amount)
```
How will this handle RGB images?

dvolgyes replied:
RGB images would be processed channel by channel. In most applications cases, this is quite fine. 
Some color space use linear combination of the RGB channels for lightness/brightness/value channel. If you convert into this space, process this channel, and convert it back, than you get approximately the same result.

Secondly, there is no perfect color space. I would not use automatic conversion inside the filters. Or if this is the decision, then make it universal everywhere in skimage. 
Maybe a decorator/wrapper could handle the conversions back and forth. 

But I think this is something what the user should do. (Does any of the function do such automatic conversion before/after processing?)

For instance, I have an RGB image, the filter converts it to HSV, process the V channel, and then?
Should it convert it back to RGB, leave it in HSV, or process it further to something else?

What about different multichannel images? E.g. 2 or 4 channels (RGBA)? If I have to choose, I would rather go for single channel version only than a multichannel with built-in color conversion, but this is just me. I would argue for consistency, and I don't see this conversion as a consistent solution across skimage. (And still: if this is the idea, then let's go for a wrapper, and do not implement it for every filter individually.)

soupault replied:
>Maybe a decorator/wrapper could handle the conversions back and forth.

@dvolgyes take a look at the discussion in https://github.com/scikit-image/scikit-image/pull/2240#issuecomment-249618562 (starting from my message).

stefanv replied:
In cases where the correct option is not clear, I agree that taking a conservative approach is good.  But, this is also an excellent opportunity to guide the user, by having the notes of the docstring describe a suggested way of handling color images.

dvolgyes replied:
So what is the conclusion here? Docstring update, or is there a general agreement about color management, like adding an option to 'convert2ycbcr', like in #2240? 

And what is the correct way of handling ndarray which have more than 3 or less than 3 color channels, but convert2ycbcr is enabled?

What is the correct way to handle color channels with 16/32bit integer types when convert2ycbcr is enabled? Without the conversion, one can process the image without issues. With conversion I have to choose a dynamic range. For 8 bit, this is 0-255, for float this is 0-1. What about 16 bit where the brightest pixel have  a pixel value of 192? 

Or should I just pass the whole image to the conversion function, and it will raise error, when it cannot do the conversion?

Finally, the code can handle arbitrary dimensions, including e.g. 4 'spatial' and a color dimension. What about this case? Should I perform an YCbCr conversion, do the unsharp masking, and convert it back? (It is a little bit of numpy indexing magic, but it can be done in a single line.) Or should the code complain about the wrong dimensions? (Or should I just pass it to the color conversion function, and it will complain anyway.)

stefanv replied:
Let's just stick to non-color images for now, and a note in the docstring on a recommended approach (convert to lab or hsv and operate on l or h).

dvolgyes replied:
Assuming that it passes the CI tests, I have no further plans in this PR. Please, let me know if you like the PR in current form, or there are parts which should be improved.

stefanv replied:
Thanks, I've added some minor tweaks (that I'm happy to also submit in the form of a PR), and mentioned a few old comments that may or may not have been addressed.

soupault replied:
I'd keep the current approach. There is absolutely no reason to drop the already implemented support for multichannel images unless it is not in line with the literature.
I'd not add `convert2ycbcr` parameter for now. Just mentioning the preferred way of using the function (pre-converting to ycbcr) should do.

emmanuelle commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,104 @@
+    return result
+
+
+def unsharp_mask(image, radius=1.0, amount=1.0, multichannel=False):
```
Was ``amount`` an obvious choice? How about ``unsharp_factor`` or ``unsharp_intensity``?

emmanuelle commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,104 @@
+
+    Parameters
+    ----------
+    image : (M[, N[, ..., P]][, C]) ndarray
```
Just ndarray would be fine, since the function is n-d (cool!).

stefanv replied:
Ping on this comment.

soupault replied:
I thought we agreed on this notation?

dvolgyes replied:
Anything is fine by me, just let me know what do you prefer in the end.

emmanuelle commented on skimage/filters/tests/test_unsharp_mask.py:
```diff
@@ -0,0 +1,41 @@
+
+
+@pytest.mark.parametrize("shape,multichannel",
+                         [((101,), False),
```
This is a huge testing matrix. Could it be less extensive so that tests run faster?

soupault replied:
Although the coverage is near to perfect here, I agree with @emmanuelle that it is better to decrease the number of tests. The main reason for this is the limitation of our free Travis CI plan: we are limited in execution time of the test suite, and having 20sec for a single function might be a bit excessive.

dvolgyes replied:
I reduced the test sizes significantly, now it should run in a few secs max.
If it is still considered slow, I could tune it a bit more.

dvolgyes commented:
Using "amount" is a natutal choice. Both textbooks and softwares (gimp,
older photoshop, etc) use radius and amount as variables. So if we use
something else, then I would put these historical names at least into the
docstring.

On Sep 1, 2017 14:10, "Emmanuelle Gouillart" <notifications@github.com>
wrote:

> *@emmanuelle* commented on this pull request.
> ------------------------------
>
> In skimage/filters/_unsharp_mask.py
> <https://github.com/scikit-image/scikit-image/pull/2772#discussion_r136559782>
> :
>
> > +    blurred = gaussian_filter(image.astype(np.float_),
> +                              sigma=radius,
> +                              mode='reflect')
> +    result = image + (image - blurred) * amount
> +
> +    # Enforce image dtype limits
> +    if image.dtype.kind == 'f':
> +        _min, _max = dtype_limits(
> +            image, clip_negative=np.all(image >= 0))
> +    else:
> +        _min, _max = dtype_limits(image, clip_negative=False)
> +    result = np.clip(result, _min, _max).astype(image.dtype)
> +    return result
> +
> +
> +def unsharp_mask(image, radius=1.0, amount=1.0, multichannel=False):
>
> Was amount an obvious choice? How about unsharp_factor or
> unsharp_intensity?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-image/scikit-image/pull/2772#pullrequestreview-60116855>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ-WK4NBQa-cBT14srByc0YD2Sw9yczks5sd_QogaJpZM4PBcDN>
> .
>

stefanv commented:
A cheap trick to get around the `amount` issue is to call it `k` or `lambda` and to describe it in the parameter section.

jni commented:
@stefanv k is in no way better than amount, and lambda is a reserved Python keyword. ;)

stefanv commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,113 @@
+
+    Returns
+    -------
+    output : (M[, N[, ..., P]][, C]) ndarray
```
Again, remove the shape information here, which is entirely general.

stefanv commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,113 @@
+    -------
+    output : (M[, N[, ..., P]][, C]) ndarray
+        Image with unsharp mask applied. It has the same shape and dtype
+        as `image`. If `image` dtype was integer, and it cannot represent
```
Why this special clipping behavior for integers, but not floats?

soupault replied:
:+1: to @stefanv 's comment

dvolgyes replied:
The input could be an uint8 type, e.g. with maximum value of 253.
The processing must happen in float format.
After the filter I get a float image which can be transformed back to the original type.
(Actually, this is the most common way.)

So in float, individual pixels could reach larger than 255 values. 
I like to use explicit clipping and not trusting overflow/underflow defaults in different languages, python included.

And the related question: what is the expected behavior in skimage? (See my other comment.) I did not understand it fully yet.

stefanv commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,113 @@
+        enhanced image = original + amount * (original - blurred)
+
+    Despite being relatively simple algorithm, it might have issues when
+    RGB color channels are processed: it might lead to color bleading.
```
processed, i.e., it might

stefanv commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,113 @@
+
+    Despite being relatively simple algorithm, it might have issues when
+    RGB color channels are processed: it might lead to color bleading.
+    Often visually more pleasing result can be achieved with processing 
```
by processing

stefanv commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,113 @@
+    RGB color channels are processed: it might lead to color bleading.
+    Often visually more pleasing result can be achieved with processing 
+    only the brightness/lightness/intensity channel in a suitable color space,
+    like HSV, HSL, YCbCr, etc. In these case, a color space conversion step
```
In this case,

stefanv commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,113 @@
+    See the skimage.color module for details about color space conversion.
+
+    Unsharp mask filter is described in most of the digital image processing books.
+    This implementation is based on Digital Image Processing[1]_.
```
`is based on [1]_`

dvolgyes commented:
I tried to address the comments but feel free to add any modification.
There is one topic I forgot, and I also don't feel it really important:
the input range for float images.
Should it be [-1,1] or [0,1]? 

I don't think so. The algorithm itself as absolutely linear, it makes no difference if we scale or shift the values. It is a bit similar to the color spaces: if there is a universal approach, then let's use it.
But if we enforce [0,1] or [-1,1], then should we enforce these for the output range too?

E.g. it is possible that someone wants radius=1 and amount=10 which will generate overshoot.
Should the user scale down the image into [-0.1,0.1] to guarantee that after USM the result will still be in [-1,1]? Or should we just ignore the output range? Or make a "clip" parameter? 

[0,1] or [-1,1]? In most color spaces there are no negative values but there are exceptions, e.g. for X-ray applications in medical imaging.

So should we enforce an arbitrary limit just for fun, give a warning (e.g. "in skimage the recommended data range is [0,1] for float-type images "), or should we just ignore this question for now, and add warning/assert when the question is settled?

I don't see any user who would use the filter and for whom it is easier to transform the image to the 0,1 range, ask the filter for clipping, and transform it back to the original range, than just clipping the result if he/she really wants it. So I think it would make the filter less convenient to use, and it doesn't help the filter implementation, accuracy, speed or any other aspect.

So I would vote for the current version, but I am open to other solutions.

stefanv commented:
We have to guarantee that floating point images in [0, 1] or [-1, 1] exit with the same range (that's how we let users build pipelines without issues).

But, we do have a keyword for handling this, called `preserve_range`, that you'll see described in the user guide, and also in `skimage.transform._warps` and `skimage.transform.filters._gaussian' e.g.

dvolgyes commented:
About output ranges:

Well, I checked them, but at least for me, the use of preserve_range is ambiguous.
"    preserve_range : bool, optional
        Whether to keep the original range of values. Otherwise, the input
        image is converted according to the conventions of `img_as_float`."

And in "img_as_float":
"The range of a floating point image is [0.0, 1.0] or [-1.0, 1.0] when converting from unsigned or signed datatypes, respectively. If the input image has a float type, intensity values are not modified and can be outside the ranges [0.0, 1.0] or [-1.0, 1.0]."

So let's check the possibilities:
- preserve_range=false, img=uint8 (max value = 192)
  -> the output can be any integer between 0 and 255

- preserve_range=false, img=float (max value = 0.75)
  the input is converted to [0,0.75]
  -> the output can be any float

- preserve_range=false, img=float (max value = 1.5)
  -> the output can be any float 

- preserve_range=true, img=uint8 (min value = 10, max value = 192)
  -> the output can be any integer between 10 and 192? or 0-255?

- preserve_range=true, img=float (min value = 0.1, max value = 0.75)
  -> the output can be any float between 0, - 1? or 0.1 , 0.75?

- preserve_range=true, img=float (min value = -0.1, max value = 1.5)
  -> the output can be any float between -0.1 and 1.5? or -1 and 1?

My interpretation is this:
- preserve_range= true: 
    - any integer type: it can use the whole integer domain, e.g. uint8 input with 192 max value, could produce result with maximum value of 255.
    - any float input between [0,1] or [-1,1] will keep this range. So if the input had only [0,0.5] range, then the output still can be [0,1]
    - any float input outside [-1,1] or [0,1] can have a range of [min(image), max(image)]???
    
- preserve_range= false: 
  - integers: the same 
  - floats: any output? regardless of its smaller or larger than 1.0?

So could someone clarify the expected behavior, including edge cases?

stefanv commented:
It means that, if you pass in values outside of the [-1, 1] or [0, 1] expected range, then the algorithm will still compute on them, without complaining.

For integers, this has no effect, since the range is fixed by the data type.

soupault commented:
> preserve_range= true:
> - any integer type: it can use the whole integer domain, e.g. uint8 input with 192 max value, could -
 produce result with maximum value of 255.
> - any float input between [0,1] or [-1,1] will keep this range. So if the input had only [0,0.5] range, then the output still can be [0,1]
> - any float input outside [-1,1] or [0,1] can have a range of [min(image), max(image)]???

> preserve_range= false:
> - integers: the same
> - floats: any output? regardless of its smaller or larger than 1.0?

I second this way of thinking. Although, it seems that for `skimage.transform.warp` ( https://github.com/scikit-image/scikit-image/blob/master/skimage/transform/_warps.py#L794 ) the meaning is slightly different.
@scikit-image/core could we have more opinions here, please?

jni commented:
@soupault that's not how preserve range was defined. Indeed, it's a lot fuzzier than this. `preserve_range` simply means that no rescaling will take place, regardless of dtype conversions. So if the algorithm requires a 16-bit image, and you give it a uint8 image, it will return a uint16 image in 0-255, and this might cause issues later on in your pipeline, for example when saving and displaying the image, which will now look very dark. (Which is why `preserve_range` always defaults to `False`). More commonly, it means that your uint8 image will be converted to a float image in 0-255 before further processing in an algorithm that depends on floating point values. But, the output range is in no way guaranteed. For example, when using an algorithm that depends on cubic interpolation (such as warp, coincidentally), you might get values higher than 255 and lower than 0.

When `preserve_range=False`, uint8 (range 0-255) will get rescaled to [0, 1] floats, [0-65335] uint16s, and so on. int8 will get rescaled to [-1, 1] floats. etc. And, clipping will take place to ensure outputs are in the predefined range for that dtype.

I guess this needs to be expanded upon in the documentation. I agree with @dvolgyes that there are many docstrings where this isn't very clear at all. At the very least, we should link to http://scikit-image.org/docs/dev/user_guide/data_types.html wherever `preserve_range` is used.

I hope that made sense, as I'm still making my way through my first morning coffee. =P

stefanv commented:
@jni That's the way I have it in my head as well.  A big +1 on linking the docs in.

dvolgyes commented:
Sorry for being away for long. So let's finalize this filter.
Theoretically, it can accept any input, even complex, and it could emit any type.
However, the high pass filter in it requires floats for meaningful precision,  
so internally, I use floats.

What about this interpretation:
- I use floats internally
- when (u)integers are emitted, they will fit into the dynamic range, regardless of the
  preserve_range keyword. If clipping is required, so be it.
- when the input is float, and preserve range is False, then the output is float, without clipping, or without any constraints on the input domain. (Could be outside of [-1,1]).
- when the input is float, and the preserve_range is True:
   - I require input from[0,1], and the output will be [0,1], using clipping if necessary.
   - if the input is not from [0,1], then the code will emit warning, clip the input for [0,1], and 
then process it as it is. (Or it could stop with assert, but I do not like this.)
- color channel: they will be processed individually (RGB or similar primary based model),
  but the docstring gets a comment about using HSV/Lab/etc color spaces where
  only the lightness/brightness/luminance/etc. channel should be processed. 
  (But I will not implement any implicit conversion.)

Is this interpretation acceptable, or do you have any other recommendation?
If you agree with these, I will send a fixed commit soon. (If you don't, then please specify your interpretation.)

stefanv commented:
I think this is a sensible interpretation.  The magical determining of [-1, 1] vs [0, 1] is a bit annoying, but I don't see any easy way around it.

dvolgyes commented:
I don't like this magic either, but I have no better idea.
Actually, I do not like this in skimage in general. 
I understand that in machine learning the [-1,1] and the [0,1] both make sense, 
but from API point of view, it would be nice to make a definite decision.
E.g. [0,1] as default, and an optional dynamic_range parameter,
 or [-1,1] as default, and bool parameter, like non_negative=True which would mean [0,1],
or something like this. 
I know that this isn't the forum to decide about this issue, 
but it would be good to clarify it on the long term.

Magic: I could default to one of the two options (e.g. [0,1]), avoid the magic, and throw a warning/error if the assumption is violated.

dvolgyes commented:
As far as I see, the PR is finished, but it fails because of the broken master branch.
Could you review it, and merge it or give feedback?
(Merge: the commit history is not really interesting here, a squash merge would be just fine.)

stefanv commented:
Another PR that's ready to go.  @scikit-image/core please review, rebase, merge as appropriate.

soupault commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,148 @@
+
+    result = image + (image - blurred) * amount
+    if vrange is not None:
+        return np.clip(image, vrange[0], vrange[1], out=result)
```
Should be clipping `result`, not `image`.

dvolgyes replied:
Good catch!

soupault commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,148 @@
+    output : ndarray
+        Image with unsharp mask applied. It returns float type, regardless
+        of the input type. The user is responsible to cast the result to
+        the expected type. If the
```
Unfinished sentence here.

dvolgyes replied:
Indeed, however, I cannot remember what was meant there. let me know, if anything is felt missing from the documentation.

soupault replied:
I think it is fine "as is".

soupault commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,148 @@
+    ill-posed problem. Because of this stability, it is often more
+    preferred than deconvolution.
+
+    The main idea is the following: Sharp details are identified as the
```
`Sharp` -> `sharp`

soupault commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,148 @@
+    should be performed before and after the unsharp masking.
+    Note that some color spaces have different than [0,1] / [0,255] dynamic range,
+    and it is the user's responsibility to take this into account.
+    See the skimage.color module for details about color space conversion.
```
Enclose `skimage.color` in \`` \``.

soupault commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,148 @@
+    ----------
+    .. [1]  Jayaraman, S., Esakkirajan, S., Veerakima, T.
+            "Digital image processing", (2009), ed iii., page 274,
+            ISBN 10: 0070144796  ISBN 13: 9780070144798
```
Add DOI, please.

dvolgyes replied:
I cannot. It is a book, it has no DOI, but ISBN.
I rather give a different book as a reference, this one has a DOI.

dvolgyes replied:
I mean, some books have DOIs, but this one did not have.

soupault replied:
Yes, sorry, my bad. Let's keep the current reference.

soupault commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,148 @@
+        if negative:
+            vrange = [-1., 1.]
+        else:
+            vrange = [0., 1.]
```
This `if-else` block is very nice and concise!

soupault commented on skimage/filters/_unsharp_mask.py:
```diff
@@ -0,0 +1,148 @@
+
+    Parameters
+    ----------
+    image : ndarray
```
It is better to specify the supported shapes of `image`.
The notation is the following: `[P, ..., ]M[, N][, C]`, i.e. the line should be `image : [P, ..., ]M[, N][, C] ndarray`. Same for `output`.

dvolgyes replied:
I do not follow the suggested notation.
What do you mean by the middle M?
In my eyes, the algorithm has the same dimension for output as the input was, and the input could be: [D1,D2,....,Dn][,C]

soupault replied:
The "cypher" from my message above is, basically, meant to explain to the user the range of supported image/ndarray shapes, and the order of different dimensions (e.g. that _channel_ dimension is the last one). The function you developed actually supports not only 2D grayscale or multichannel images, but also 1D grayscale/multichannel arrays and 3D+. This information could come very useful to the users.
Currently (at least, as I perceive it), `M` fuzzily stands for _rows_, `N` for _columns_, `C` for _channels_, and `P, etc` for any extra spatial dimensions.
I'd highly recommend you to take a look at the discussions in https://github.com/scikit-image/scikit-image/issues/2276, https://github.com/scikit-image/scikit-image/pull/3031.

soupault commented on skimage/filters/tests/test_unsharp_mask.py:
```diff
@@ -0,0 +1,69 @@
+        array /= max(np.abs(array).max(), 1.0)
+    output = unsharp_mask(array, radius, amount, multichannel, preserve)
+    assert output.dtype in [np.float32, np.float64]
+    assert output.shape == shape
```
How long does the testing take?

dvolgyes replied:
A few seconds, not too fast, not too slow. The operations themselves are cheap, and the problem size is small, so it should not run for long. But there are a few variations, so it adds up.

soupault replied:
Ok, fine for this. :) I was worried about our utilization of the CI, but couple of seconds are well affordable.

soupault commented on skimage/filters/tests/test_unsharp_mask.py:
```diff
@@ -0,0 +1,69 @@
+import pytest
+
+
+@pytest.mark.parametrize("shape,multichannel",
```
Also, please, do not use `pytest` directly. Instead, import `parametrize` from `skimage._shared.testing` (our thin wrapper) (as in https://github.com/scikit-image/scikit-image/blob/master/skimage/util/tests/test_dtype.py, for example).

dvolgyes commented:
I think I addressed most issues, except the image dimension part where I do not understand the suggestion. If you could detail it, I would appreciate it. Or if/when I figure it out on my own, I will commit the fix.

dvolgyes commented:
Thanks for the review, the comments, and the cosmetic changes. I assume the PR will be closed by someone from skimage when it is actually merged, so I do not plan to do anything. If anything is required from my side, just let me know.

stefanv commented:
Merged in https://github.com/scikit-image/scikit-image/commit/372c393176af42443a64eddd5f44f0547addcfad
